src/rgw/rgw_sal_motr.cc

```diff
@@ -3419,55 +3356,75 @@ int MotrAtomicWriter::complete(size_t accounted_size, const std::string& etag,
-	// TODO: This will be taken care during version implementation.
-    	rc = old_mobj->delete_object(dpp, NULL, y, true);
-	if (rc == 0) {
-      		ldpp_dout(dpp, 20) <<__func__<< ": Old " << obj_type << " ["
-          	<< old_mobj->get_name() <<  "] deleted succesfully" << dendl;
-    	} else {
-      		ldpp_dout(dpp, 0) << __func__<<": Failed to delete old " << obj_type << " ["
-          	<< old_mobj->get_name() <<  "]. Error = " << rc << dendl;
-                // TODO: This will be handled during GC
-    	}
-  	}
-  }
   rc = store->do_idx_op_by_name(bucket_index_iname,
                                 M0_IC_PUT, obj.get_key().to_str(), bl);
-
   if (rc != 0) {
     // TODO: handle this object leak via gc.  
     ldpp_dout(dpp, 0) << __func__ << ": index operation failed, rc = "<< rc << dendl;
     return rc;
   }
-
   store->get_obj_meta_cache()->put(dpp, obj.get_key().to_str(), bl);
 
+  if( !info.versioning_enabled() )
+  {
+    std::string obj_type = "simple object";
+    rgw_bucket_dir_entry old_ent;
+    bufferlist old_check_bl;
+    rc = obj.fetch_null_obj(dpp, old_check_bl);
+    if (rc < 0){
+      ldpp_dout(dpp, 0) <<__func__<< ": Failed to fetch null object, rc : " << rc << dendl;
+      return rc;
+    }
+    // TODO: Remove this call in optimization tkt CORTX-31977
+    std::string prev_obj_key;
+    rc = obj.fetch_null_obj_reference(dpp, prev_obj_key);
+    if (rc < 0){
+      ldpp_dout(dpp, 0) <<__func__<< ": Failed to fetch null reference key, rc : " << rc << dendl;
+      return rc;
+    }
+    rc = obj.update_null_reference(dpp, ent);
+    if (rc < 0){
+      ldpp_dout(dpp, 0) <<__func__<< ": Failed to update null reference key, rc : " << rc << dendl;
+      return rc;
+    }
+    // error log handled in above function.
+	  // Before updating bucket index with entry for new object, check if
+  	// old object exists.
+  	if (rc == 0 && old_check_bl.length() > 0) {
+      std::unique_ptr<rgw::sal::Object> old_obj = obj.get_bucket()->get_object(rgw_obj_key(obj.get_name()));
+      rgw::sal::MotrObject *old_mobj = static_cast<rgw::sal::MotrObject *>(old_obj.get());
+      auto ent_iter = old_check_bl.cbegin();
+      old_ent.decode(ent_iter);
+      rgw::sal::Attrs attrs;
+      decode(attrs, ent_iter);
+      old_mobj->meta.decode(ent_iter);
+      std::string prev_obj_instance = old_ent.key.instance;
+      old_mobj->set_instance(prev_obj_instance);
+      if (old_ent.meta.category == RGWObjCategory::MultiMeta) {
+          obj_type = "multipart object";
+          old_mobj->set_category(RGWObjCategory::MultiMeta);
+      }
+      ldpp_dout(dpp, 20) <<__func__<< ": Old " << obj_type << " exists" << dendl;
+      // Delete old object
+      // Initialize version of the previous object so that it can delete a specific version,rather than delete the latest object.
+      // TODO: This will be taken care during version implementation.
+      rc = old_mobj->remove_mobj_and_index_entry(dpp, old_ent, prev_obj_key);
```

Comment:
- Before calling remove_mobj_and_index_entry() function we have PUT the new obj, fetch_null_refernce key (previous null obj) and updated null reference key(obj/null). 

- The key that we have fetched in fetch_null_reference call that key we are passing to remove_mobj_and_index_entry() this function. If we do not pass the key here and fetch the key in this function we may need to call fetch_null_reference() from multiple places (ie complete() and remove_mobj_and_index_entry()).

- This can be optimize in another tkt [CORTX-31977](https://jts.seagate.com/browse/CORTX-31977).